[AIRFLOW-1011] Fix bug in BackfillJob._execute() for SubDAGs

joeschmid opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issue:

- https://issues.apache.org/jira/browse/AIRFLOW-1011

Testing Done:

- Unit tests pass on Travis

BackfillJob._execute() checks that the next run date is less than
or equal to the end date before creating a DAG run and task
instances. For SubDAGs, the next run date is not relevant,
i.e. schedule_interval can be anything other than None
or '@once' and should be ignored. However, current code calculates
the next run date for a SubDAG and the condition check mentioned
above always fails for a SubDAG triggered manually, hence the tasks
in the SubDAG will never be executed.

This change adds a simple check to determine if this is a SubDAG
and, if so, sets next run date to DAG run's start date.

bolkedebruin commented on airflow/jobs.py:
```diff
@@ -1774,8 +1774,11 @@ def _execute(self):
+        # next run date for a subdag isn't relevant (schedule_interval for subdags
+        # is ignored) so we use the dag run's start date in the case of a subdag
+        next_run_date = self.dag.normalize_schedule(dr_start_date) \
+            if "." not in self.dag.dag_id else dr_start_date
```
please use () across multiple lines and use "dag.is_subdag" instead of manually checking

joeschmid replied:
Thanks, Bolke! Just added the () across multiple lines, but it looks like dag.is_subdag is not available everywhere. (Got "AttributeError: 'DAG' object has no attribute 'is_subdag'" when I tried it there.) Also, I'm working on adding unit tests now.

bolkedebruin replied:
Have a look at:

https://github.com/apache/incubator-airflow/blob/master/airflow/models.py#L201

I think it is actually a bug that it is not properly exposed. Can you please add it to the "DAG class", I think that to be a more proper fix.

bolkedebruin reviewed (approved):
I consider this a workaround the actual solution is dat a subdag should just "have" the schedule of its parent.

please add unit tests (not integration tests|)

joeschmid committed:
[AIRFLOW-1011] Fix bug in BackfillJob._execute() for SubDAGs

BackfillJob._execute() checks that the next run date is less than
or equal to the end date before creating a DAG run and task
instances. For SubDAGs, the next run date is not relevant,
i.e. schedule_interval can be anything other than None
or '@once' and should be ignored. However, current code calculates
the next run date for a SubDAG and the condition check mentioned
above always fails for SubDAG triggered manually.

This change adds a simple check to determine if this is a SubDAG
and, if so, sets next run date to DAG run's start date.
